node/subsystem-util/src/memvisor.rs

```diff
@@ -0,0 +1,166 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Polkadot Memvisor helper for tracing memory usage across subsystems.
+
+#![forbid(unused_imports)]
+
+use metrics::Metrics;
+use std::{default::Default, sync::Arc};
+
+use crate::{
+	metrics,
+	metrics::{
+		prometheus,
+		prometheus::{CounterVec, Opts, PrometheusError, Registry, U64},
+	},
+};
+
+#[derive(Default, Clone)]
+#[cfg_attr(test, derive(Debug))]
+pub struct MemVisorMetrics(Option<MemVisorMetricsInner>);
+
+#[derive(Clone)]
+#[cfg_attr(test, derive(Debug))]
+pub struct MemVisorMetricsInner {
+	/// Counters for memory (de)allocation.
+	/// Labels:
+	/// - `span_name`
+	/// - `object` ... the obejct we are tracking memory for.
+	mem_alloc: CounterVec<U64>,
+	mem_free: CounterVec<U64>,
+}
+
+impl Metrics for MemVisorMetrics {
+	fn try_register(registry: &Registry) -> Result<Self, PrometheusError> {
+		let metrics = MemVisorMetricsInner {
+			mem_alloc: prometheus::register(
+				CounterVec::new(
+					Opts::new("memvisor_alloc", "Memory footprint increase for a span."),
+					&["span_name", "object"],
+				)?,
+				registry,
+			)?,
+			mem_free: prometheus::register(
+				CounterVec::new(
+					Opts::new("memvisor_free", "Memory footprint decrease for a span."),
+					&["span_name", "object"],
+				)?,
+				registry,
+			)?,
+		};
+		Ok(MemVisorMetrics(Some(metrics)))
+	}
+}
+
+#[derive(Clone)]
+#[cfg_attr(test, derive(Debug))]
+pub struct MemSpan {
+	/// The name of the span.
+	name: &'static str,
+	/// The metric value we are tracking.
+	value: u64,
+	/// Reference to MemVisorMetrics to enable metric reporting
+	/// from inside spans.
+	visor_metrics: Arc<MemVisorMetrics>,
+}
+
+#[derive(PartialEq)]
+#[cfg_attr(test, derive(Debug))]
+enum MemSpanEvent {
+	None,
+	Alloc(u64),
+	Free(u64),
+}
+
+impl MemSpan {
+	pub fn new(name: &'static str, visor_metrics: Arc<MemVisorMetrics>) -> Self {
+		MemSpan { name, value: 0, visor_metrics }
+	}
+
+	pub fn child(&self, name: &'static str) -> MemSpan {
+		MemSpan { name, value: 0, visor_metrics: self.visor_metrics.clone() }
+	}
+
+	/// Enable monitoring for a discrete list of suspect data structures.
+	/// The total usage should be recorded with this call
+	pub fn start(&mut self, start_value: u64) {
+		if self.visor_metrics.0.is_some() {
+			self.value = start_value;
+		}
+	}
+
+	/// Ends current span and update metrics.
+	pub fn end(&mut self, stop_value: u64) {
```

Comment:
I'd prefer to make this implicit by a `Drop`.

Reply:
On second thought, this might be not possible, since that would imply that a reference to the type has to be held across changing it, which would violate rust's aliasing rules.

One workaround would involve a macro that does both of these 

```
macro_rules! mem_visor {
   ($name:lit, $parent:ident, $track_me:ident, $action: expr) => { let span = parent.child($name); span.start($track_me); let val = $action; span.end($track_me); val }
}
```

or similar... just food for thought though